runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -2876,8 +2876,14 @@ J9::ValuePropagation::getParmValues()
                   }
                }
             }
-         else if (0) //added here since an unresolved parm could be an interface in which case nothing is known
+         else
             {
+            int32_t len;
+            const char *sig = p->getTypeSignature(len);
+            constraint = createTypeHintConstraint(method, sig, len);
```

Comment:
Similarly to `innerConstrainAcall()`, could you please create the hint also when there is a type that has been erased due to an interface type? If you pick up the changes from #16316, then I think you'll find that catching that case is necessary to find a hint when the parameter is declared to be an array of an interface type

Reply:
I experimented a bit. I think the code should be changed as below. The check on `if (!opaqueClass)` will catch the case when the `opaqueClass` is erased by `isUnreliableSignatureType`. 

**=== OLD ===**
```c++
J9::ValuePropagation::getParmValues()
   {
...
         if (opaqueClass)
            {
...
               TR_OpaqueClassBlock *erased = NULL;
               if (isUnreliableSignatureType(opaqueClass, erased))
                  opaqueClass = erased;
...
            }
        else
            {
            int32_t len;
            const char *sig = p->getTypeSignature(len);
            constraint = createTypeHintConstraint(method, sig, len);
            }
```

**=== NEW ===**
```c++
J9::ValuePropagation::getParmValues()
   {
...
         if (opaqueClass)
            {
...
               TR_OpaqueClassBlock *erased = NULL;
               if (isUnreliableSignatureType(opaqueClass, erased))
                  opaqueClass = erased;
...
            }

         if (!opaqueClass)
            {
            int32_t len;
            const char *sig = p->getTypeSignature(len);
            TR::VPConstraint *typeHintConstraint = createTypeHintConstraint(method, sig, len);

            if (typeHintConstraint)
               constraint = constraint ? constraint->intersect(typeHintConstraint, this) : typeHintConstraint;
            }
```

Reply:
`erased` can be non-null

Reply:
Right, I'll add a flag as below and check if the class has been erased. 
```c++
J9::ValuePropagation::getParmValues()
   {
...
         bool isClassErased = false;
         if (opaqueClass)
            {
...
               TR_OpaqueClassBlock *erased = NULL;
               if (isUnreliableSignatureType(opaqueClass, erased))
                  {
                  isClassErased = true;
                  opaqueClass = erased;
                  }
...
            }

         if (!opaqueClass || isClassErased)
            {
            int32_t len;
            const char *sig = p->getTypeSignature(len);
            TR::VPConstraint *typeHintConstraint = createTypeHintConstraint(method, sig, len);

            if (typeHintConstraint)
               constraint = constraint ? constraint->intersect(typeHintConstraint, this) : typeHintConstraint;
            }
```

Reply:
Addressed in 53864c1